### DISCUSSION TITLE
Feature Request: range types

### DISCUSSION THREAD

**Discussion by snth** [duckdb/duckdb#6077] [Ideas]
Hi,

### Why do you want this feature?

I'm a big fan of DuckDB and I believe the following feature would make my favourite database quack even louder.

We're big users of the [postgres range types](https://www.postgresql.org/docs/current/rangetypes.html) and [range operators](https://www.postgresql.org/docs/current/functions-range.html). I think they are a pretty great feature. They greatly simplify queries looking for overlaps of date ranges etc... and I wouldn't want to have to give those up when shifting my workloads over to DuckDB.

I believe you can do the same with (start, end) pairs but writing `WHERE (a.start <= b.start AND a.end >= b.start) OR (reverse case)` are a pain to write and easy to get wrong. It's much easier and safer to write `WHERE a.range && b.range`. 

Postgres 14 also introduced [range_agg and range_intersect_agg aggregation functions](https://www.postgresql.org/docs/current/functions-aggregate.html) which make some new use cases (or at least simple) which would have previously required complex recursive CTEs.

Overall I believe that range types would be a great Quality Of Life improvement for users and would make DuckDB even more quacking awesome!

### References

Postgres range types: https://www.postgresql.org/docs/current/rangetypes.html
Postgres range functions and operators: https://www.postgresql.org/docs/current/functions-range.html
Postgres range aggregations: https://www.postgresql.org/docs/current/functions-aggregate.html

**Comment by Maxxen** [comment]
@Tishj I think this is a good intermediate-level project for some of our new interns no?

**Comment by snth** [comment]
Just following up on this whether you have given this any more thought?

IMHO this is one of the coolest features in PostgreSQL and I use it heavily. It's great for finding gaps in data and the like, especially since the introduction of `multirange`s. Look for example at the following simple example:

```sql
with data as (
    select 
        daterange('2024-04-01', '2025-03-31') AS financial_year,
        daterange('2024-06-20', '2024-06-30') AS closed_period_1,
        daterange('2024-12-20', '2024-12-31') AS closed_period_2
),
calendar as (
    select *, multirange(financial_year) - multirange(closed_period_1) - multirange(closed_period_2) as trading_periods
    from data
)
select unnest(trading_periods) AS trading_period
from calendar
```
output:
|trading_period|
|-|
|[2024-04-01,2024-06-20)|
|[2024-06-30,2024-12-20)|
|[2024-12-31,2025-03-31)|

It would be really great to have access to this in DuckDB as well!

Some references:
* [9.20. Range/Multirange Functions and Operators (PostgreSQL Documentation)](https://www.postgresql.org/docs/current/functions-range.html)
* [HN: Multiranges in Postgres (Oct 2022)](https://news.ycombinator.com/item?id=33353719)
  * Some more testimonials about the usefulness of this feature in the comments.

**Reply by Maxxen** [reply]
Now that we have the fixed-size-array type, would it make sense to implement some of these functions on top of that, e.g treating any fixed-size array of size 2 as a "range"?. Alternatively, on top of structs with min/max members?

**Reply by snth** [reply]
Thanks for the quick response.

I think there's more to it than that. First of all the ranges also have a 3rd parameter in the constructor about whether bounds are included, so you have `[]`,`()`,`[)`, and `(]` types.

Furthermore the main thing that makes them nice is all the operators that are implemented on top of them, so things like intersection, union, etc... and very importantly for my use cases the `range_agg` aggregation operator which can return a `multirange`.

I'm impartial as to what internal representation you pick as it's the functionality that I'm really after. The binary operators I could probably work around with (messy) SQL but the aggregation and multirange operations allow use cases that I can't picture doing easily otherwise.

**Reply by cstork** [reply]
Just to support @snth's reply: especially multiranges are the secrete sauce that make some applications feasible that would be hell to implement without. With respect to DuckDB, ranges allow for simpler and more general analytical queries — especially of temporal data.

**Reply by snth** [reply]
@Maxxen You folks are the wizards of writing efficient analytics code so I'm sure you will have much more insight on the implementation side than me. From what I can tell, for `range`s you would need a 2-element tuple/struct with min/max members as you pointed out as well as an additional 2-bits for the endpoint in/ex-clusions.

For `multirange`s I expect you need some sort of balanced search tree to use as the accumulator for efficient lookups during aggregations.

TIL: [Interval Tree](https://en.wikipedia.org/wiki/Interval_tree) This might be useful for creating an index at the column level for a `range` column. For the `multirange` accumulator during a `range_agg` it probably does more than you need since the interval tree allows for overlapping ranges while I would maintain the multirange accumulator as non-overlapping ranges which should allow additional optimisation.

Relevant portions of the PostgreSQL implementation look like they could be found in [multirangetypes.c](https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/multirangetypes.c). I just had a quick look and it looks like they just accumulate in an array and then sort and combine. The [multirange_intersect_internal](https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/multirangetypes.c#L1260) code also looks interesting.

**Reply by Maxxen** [reply]
This does sound like a really fun thing to implement but unfortunately I don't expect we will be able to tackle this properly before next release. However, im thinking representing ranges as a `STRUCT<T> (min <T>, max <T>, endpoints_type UTINYINT)` would work pretty well, ill see if I can find some time to play around with this later.

**Comment by soerenwolfers** [comment]
+1 for ranges. Beyond that, I think it would be great if DuckDB took the opportunity and offer something even better than multiranges when they get to this: (step) functions.

--- 

A multirange is simply a step function with arbitrary domain and boolean co-domain (where set containment becomes function evaluation). For the purpose of this post I will denote this via `multirange<T> = StepFunction<T, BOOL>`.

Step functions with other co-domains, `StepFunction<T, S>`, allow for even more applications. For example:

(0) A `multirange<TIME> = StepFunction<TIME, BOOL>` can encode whether someone is busy or not.
(1) A `StepFunction<TIME, ENUM / VARCHAR>` can encode *what* they're busy with.
(2) A population timeseries or a ranking timeseries can be a `StepFunction<DATE, INT>`
(3) A price timeseries can be a `StepFunction<TIMESTAMP, FLOAT>`
(4) A marginal income tax rate system could be a `StepFunction<FLOAT, FLOAT>`
(6) A strategy / policy can be a `StepFunction<FLOAT, ENUM>` 
(7) A discount function can be a `StepFunction<INT, FLOAT>` (percentage discount for `n` units bought)

Some of these might be stored directly, others may come up naturally in the analytic manipulation of more basic in- and output data. Examples:

**Calendars**

If you have columns `employeeName` (`VARCHAR`) and `schedule` (`StepFunction<TIME, VARCHAR>`) and want to know what your employees are currently up to, you can do

```sql
SELECT
    employeeName, 
    schedule.At(current_time) -- VARCHAR
```

If you want to know the time where they're all available, you can do

```sql
SELECT
    StepFunctionAggregate('bool_and', schedule.Apply(lambda x: x == 'idle') -- multirange<TIME>
```

If there is no such time and you want to find the time where the most employees are available, you can do

```sql
SELECT
    numbersAvailable: StepFunctionAggregate('sum', available::StepFunction<TIME, INT>), -- StepFunction<TIME, INT>
    maximalNumberAvailable: numbersAvailable.Max(), -- INT
    mostAvailable: numbersAvailable.Apply(lambda x: x == maximalNumberAvailable), -- multirange<TIME>
    firstMostAvailable: mostAvailable.Leftmost() -- TIME
```

(where I have used `available` as a shorthand for the previously defined `schedule.Apply(lambda x: x == 'idle')`).

**Populations**

If you have a column `population` (`StepFunction<DATE, INT>`) and want to know the first time the population was above `10_000`, you can do 

```sql
SELECT
    population.Apply(lambda x: x > 10_000).Leftmost()
```

If you want to know the first time it was _continuously_ above `10_000` for five years, you can do

```sql
SELECT
    population.Window(INTERVAL (-5) YEARS, INTERVAL 0 YEARS, 'min').Apply(lambda x: x > 10_000).Leftmost()
```

**Prices**

 If you have a column `prices` (`StepFunction<TIMESTAMP, FLOAT>`) and want to know the _n_th time the price crossed the 100$ barrier (and you know the price started out below 100$), you can do

```sql
SELECT
    pieces: prices.Apply(lambda x: x > 100).Pieces(), -- List[(x Range, y bool)]
    nthCrossing: pieces[2 * n].x.left
```


If you have two columns `pricesA` and `pricesB` and you have convinced yourself that the two corresponding stocks always trade at the same price in the long-term, then trading opportunities (i.e., times where the difference of the two prices is larger than the cost to sell one and buy the other and later unwind that trade) can be found via  

```sql
SELECT
    priceDifference: StepFunction.Apply(stockAPrices, stockBPrices, lambda a, b: a - b), -- StepFunction<TIMESTAMP, FLOAT>
    tradingOpportunities: priceDifference.Apply(cost, lambda d, c: abs(d) > c) -- mutirange<bool>
```
    


**Incomes**

If you have a variable `grossIncome` (`float`) and a variable `marginalTaxRates` (`StepFunction<FLOAT, FLOAT>`), then `grossIncome` is simply 

```sql
SELECT 
    marginalTaxRates.Integral(0, grossIncome)
```

---

A lot of the queries shown can be performed using sparse "long" data and `ASOF` joins instead, but that's way more tricky, comes with fewer automatic sanity checks (e.g.: do start and end dates alternate, is the ordering in my query correct), and is likely slower because evaluations, apply-cations, and all other analytics on `StepFunction` (e.g., `contains` / `intersect` / `union` / `max` / `leftmost` / `rightmost` / `integral` / `window`)  have trivial and super fast implementations.


(I challenge anybody to write the "find best time for a meeting" query above in regular SQL.)
